packages/core/src/queue/event-queue.ts

```diff
@@ -0,0 +1,319 @@
+    }
+
+    return false
+  }
+
+  async flush(): Promise<CoreContext[]> {
+    if (this.queue.length === 0 || isOffline()) {
+      return []
+    }
+
+    let ctx = this.queue.pop()
+    if (!ctx) {
+      return []
+    }
+
+    ctx.attempts = this.queue.getAttempts(ctx)
+
+    try {
+      ctx = await this.deliver(ctx)
+      this.emit('flush', ctx, true)
+    } catch (err: any) {
+      const accepted = this.enqueuRetry(err, ctx)
+
+      if (!accepted) {
+        ctx.setFailedDelivery({ reason: err })
+        this.emit('flush', ctx, false)
+      }
+
+      return []
+    }
+
+    return [ctx]
+  }
+
+  private isReady(): boolean {
+    // return this.plugins.every((p) => p.isLoaded())
+    // should we wait for every plugin to load?
+    return true
+  }
+
+  private availableExtensions(denyList: Integrations): CorePluginsByType {
+    const available = this.plugins.filter((p) => {
+      // Only filter out destination plugins or the Segment.io plugin
+      if (p.type !== 'destination' && p.name !== 'Segment.io') {
+        return true
+      }
+
+      // Explicit integration option takes precedence, `All: false` does not apply to Segment.io
+      return (
+        denyList[p.name] ??
+        (p.name === 'Segment.io' ? true : denyList.All) !== false
+      )
+    })
+
+    const {
+      before = [],
+      enrichment = [],
+      destination = [],
+      after = [],
+    } = groupBy(available, 'type')
+
+    return {
+      before,
+      enrichment,
+      destinations: destination,
+      after,
+    }
+  }
+
+  private async flushOne(ctx: CoreContext): Promise<CoreContext> {
+    if (!this.isReady()) {
+      throw new Error('Not ready')
+    }
+
+    const { before, enrichment } = this.availableExtensions(
+      ctx.event.integrations ?? {}
+    )
+
+    for (const beforeWare of before) {
+      const temp: CoreContext | undefined = await ensure(ctx, beforeWare)
+      if (isContext(temp)) {
+        ctx = temp
+      }
+    }
+
+    for (const enrichmentWare of enrichment) {
+      const temp = await attempt(ctx, enrichmentWare)
+      if (isContext(temp)) {
+        ctx = temp
+      }
+    }
+
+    this.emit('message_enriched', ctx) // TODO: inspectorHost.enriched?.(ctx as any)
+
+    // Enrichment and before plugins can re-arrange the deny list dynamically
+    // so we need to pluck them at the end
+    const { destinations, after } = this.availableExtensions(
+      ctx.event.integrations ?? {}
+    )
+
+    await new Promise((resolve, reject) => {
+      setTimeout(() => {
+        const attempts = destinations.map((destination) =>
+          attempt(ctx, destination)
+        )
+        Promise.all(attempts).then(resolve).catch(reject)
+      }, 0)
+    })
+
+    this.emit('message_delivered', ctx)
+    ctx.stats?.increment('message_delivered')
+    // should emit to: TODO: inspectorHost.delivered?.(ctx as any, ['segment.io']) -- can we move all this inspector stuff to a plugin that listens to these events?
```

Comment:
don't want inspectorHost in core -- it can use the event emitter to listen to events

Reply:
Sounds like a `utility` plugin - /cc @zikaari